lib/collection/src/shards/replica_set/mod.rs

```diff
@@ -568,6 +568,11 @@ impl ShardReplicaSet {
                             .await?;
                         self.notify_peer_failure(peer_id);
                     }
+                    ReplicaState::PartialSnapshot => {
+                        self.set_local(local_shard, Some(ReplicaState::PartialSnapshot))
+                            .await?;
+                        self.notify_peer_failure(peer_id); // TODO!? 🤔
+                    }
```

Comment:
Do you mean that we can subscribe on this signal as a mechanism for the sender to wait for consensus to be propagated?

I don't think this is required. I believe the sender has to wait on all nodes explicitly for consensus to be propagated, in which case it is already waiting on this node as well. Then we better use this notification system for its intended purpose.

Reply:
I don't get it. This whole method is called when *collection* snapshot is being recovered. And if we got a shard in `PartialSnapshot` state after recovering collection snapshot, that simply means that collection snapshot was made while shard snapshot transfer was in progress. And so, we have to *somehow* switch this replica from `PartialSnapshot` into some valid state.

Reply:
I mean the call to `notify_peer_failure` specifically.

Reply:
Still don't understand it. But I'll try to explain what's going on a bit better?

`notify_peer_failure` simply reports to the consensus to switch the node into `Dead` state (here's where [`notify_peer_failure_cb` is defined][notify-peer-failure]; it's propagated from `TableOfContent::create_collection` down to the `ShardReplicaSet`). If we simply ignore it here, then the node would be stuck in `PartialSnapshot` state indefinitely, so we either have to handle it here somehow, or somewhere else. (In `sync_local_state`, maybe?)

Though, I'm not sure what's the best way to handle it. E.g., what would happen if we try to recover a collection snapshot that contains a node in `PartialSnapshot` state, and specify `SnapshotPriority::Snapshot`? Would it break all replicas of a particular shard? 🤔

I think we need @generall input on this one.

Reply:
I think I figured it out... and turned out I was confused about `apply_state`. :/

So, `apply_state` is called when _raft state snapshot_ is applied to the node. (Currently, this can only be triggered with explicit `/cluster/recover` request.)

State snapshot transfer is organized and handled transparently by the `raft` crate, and `apply_state` is the method that should actually apply collection data from the snapshot to the local node.

This particular part of `apply_state` updates the state of the _local_ shard on the node. But, considering that this is _raft state_ snapshot:

- this state can always be considered "fresh"
  - the snapshot is sent by the Raft leader
  - and transmission is organized as part of Raft message exchange
    - (so `raft` should handle delays and transmission errors)
- so, if there's a shard in `Partial` or `PartialSnapshot` state
  - either there's a shard transfer in progress currently (or it is about to begin, or it just ended)
  - or we would receive Raft update that would switch this node back to `Dead` shortly

So, we should simply apply the state as-is and move on. No need to `notify_peer_failure` or anything else.